lib/galaxy/jobs/runners/drmaauniva.py

```diff
@@ -0,0 +1,640 @@
+            return self.drmaa.JobState.FAILED
+
+        state = self.drmaa.JobState.UNDETERMINED
+        # exit_status
+        # Exit status of the job script (or Univa Grid Engine specific status in case of certain
+        # error conditions). The exit status is determined by following the normal shell conventions
+        # If the command terminates normally the value of the command is its exit status.
+        # However, in the case that the command exits abnormally, a value of 0200 (octal), 128
+        # (decimal) is added to the value of the command to make up the exit status.
+        # For example: If a job dies through signal 9 (SIGKILL) then the exit status
+        # becomes 128 + 9 = 137.
+        if "exit_status" in qacct:
+            qacct["exit_status"] = int(qacct["exit_status"])
+            if qacct["exit_status"] < 1:
+                logging.error( "DRMAAUniva: job {job_id} has exit status {status}".format(job_id=job_id, status=qacct["exit_status"]) )
+                state = self.drmaa.JobState.DONE
+            elif 0 < qacct["exit_status"] < 129:
+                logging.error( "DRMAAUniva: job {job_id} has exit status {status}".format(job_id=job_id, status=qacct["exit_status"]) )
+                extinfo[ 'exit_status' ] = qacct["exit_status"]
+                state = self.drmaa.JobState.FAILED
+            else:
+                logging.error( "DRMAAUniva: job {job_id} was killed by signal {signal}".format(job_id=job_id, signal=qacct["exit_status"] - 128) )
+                state = self.drmaa.JobState.FAILED
+                extinfo[ "signal" ] = signals[ qacct["exit_status"] - 128 ]
+
+        # failed
+        # Indicates the problem which occurred in case a job could not be started on the execution
+        # host (e.g. because the owner of the job did not have a valid account on that
+        # machine). If Univa Grid Engine tries to start a job multiple times, this may lead to
+        # multiple entries in the accounting file corresponding to the same job ID.
+        # for the codes see https://docs.oracle.com/cd/E19957-01/820-0699/chp11-2/index.html
+        if "failed" in qacct:
+            code = int(qacct["failed"].split()[0])
+            # this happens in case of no error or exit_code!=0 (0) or a signal (100).
+            # both cases are covered already
+            if code in [ 0, 100 ]:
+                pass
+            # these seem to be OK to
+            elif code in [ 24, 25 ]:
+                state = self.drmaa.JobState.RUNNING
+            else:
+                logging.error( "DRMAAUniva: job {job_id} failed with failure {failure}".format(job_id=job_id, failure=qacct["failed"]) )
+                state = self.drmaa.JobState.FAILED
+        logging.debug("DRMAAUnivaJobRunner._get_drmaa_state_qacct ({jobid}) -> {state}".format(jobid=job_id, state=self.drmaa_job_state_strings[state]))
+        return state
+
+    def _get_drmaa_state_wait( self, job_id, ds, extinfo ):
```

Comment:
This could move up to the parent class since it's DRM-agnostic.

Reply:
Could be done, but does this make sense if no other runner uses it? What would be the rationale?